src/qilib/data_set/mongo_data_set_io_reader.py

```diff
@@ -0,0 +1,213 @@
+            adjusted_updates = self._convert_dot_notation_to_dict(updated_fields)
+            self._update_data_set(adjusted_updates)
+            empty_queue = self._update_queue.empty()
+
+    def _convert_dot_notation_to_dict(self, updated_fields: Dict[str, Any]) -> Dict[str, Any]:
+        """ If a nested field is updated in the mongo database, the nested document is replace with a dot notation that
+            does not match the data set that is to be updated.
+            For example:
+                {data_arrays.array_name: {<data>}}
+            Needs to be converted to:
+                {data_arrays: {array_name: {<data>}}
+
+        Args:
+            updated_fields: An update event from the mongo database change stream.
+
+        Returns:
+            An updated dictionary without the dot notation.
+
+        """
+        adjusted_updates: Dict[str, Any] = {}
+        for update in updated_fields.keys():
+            key = update.split('.')
+            if len(key) > 1 and key[1].isnumeric():
+                # {array_updates.index: [<index>, <data>] -> {array_updates: [[<index>, <data>]]
+                adjusted_updates[key[0]] = [updated_fields[update]]
+            elif len(key) > 1 and (key[0] == self.DATA_ARRAYS or key[0] == self.METADATA):
+                # {data_arrays.array_name: {<data>}} -> {data_arrays: {array_name: {<data>}}
+                adjusted_updates[key[0]] = {key[1]: updated_fields[update]}
+            else:
+                adjusted_updates[key[0]] = updated_fields[update]
+        return adjusted_updates
+
+    def bind_data_set(self, data_set: DataSet) -> None:
+        """ Binds the DataSet to the DataSetIOReader.
+
+        Args:
+            data_set: The object that encompasses DataArrays.
+
+        """
+        self._data_set = data_set
+        document = self._mongo_data_set_io.get_document()
+        self._data_set.name = document.get('name')
+        self._update_data_set(document)
+
+    def _update_data_set(self, document: Any) -> None:
+
+        if self.METADATA in document:
+            for field, value in document.get(self.METADATA).items():
+                setattr(self._data_set, field, value)
+        if self.DATA_ARRAYS in document:
+            set_arrays = list(filter(lambda a: a['is_setpoint'], document.get(self.DATA_ARRAYS).values()))
+            self._update_set_arrays(set_arrays)
+            data_arrays = list(filter(lambda a: not a['is_setpoint'], document.get(self.DATA_ARRAYS).values()))
+            for array in data_arrays:
+                if hasattr(self._data_set, array['name']):
+                    self._update_data_array(array)
+                else:
+                    self._data_set.add_array(self._construct_data_array(array))
+        if self.ARRAY_UPDATES in document:
+            for array_update in document.get(self.ARRAY_UPDATES):
+                index_or_slice = tuple(array_update[0])
```

Comment:
If I do this in the producer:

`data_set.add_data((1), {'z': [1,2]})`

I get this in the consumer:

```
>>> consumer_dataset.sync_from_storage(-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kfjmartens/src/generic/qilib/src/qilib/data_set/data_set.py", line 138, in sync_from_storage
    self._storage_reader.sync_from_storage(timeout)
  File "/home/kfjmartens/src/generic/qilib/src/qilib/data_set/mongo_data_set_io_reader.py", line 98, in sync_from_storage
    self._update_data_set(adjusted_updates)
  File "/home/kfjmartens/src/generic/qilib/src/qilib/data_set/mongo_data_set_io_reader.py", line 157, in _update_data_set
    index_or_slice = tuple(array_update[0])
TypeError: 'int' object is not iterable
>>>
```

Reply:
that´s not good. But fixed.